### ISSUE TITLE
Updates to pooling script in Twist WF

### ISSUE THREAD

**Issue by annagellerbring** [Clinical-Genomics/clinical_EPPs#149]
Lims Issue
---------------

**Work Flow:** TWIST v2
**Protocol:** Target Enrichment TWIST v1
**Step:** Pool samples for hybridization TWIST

**Description:** Updating script for calculating volumes to pool with requirements:
- minimum input amount to pool: 187.5ng
- if input amount to pool is <187.5ng give error message that explains that the pool contains too many samples, and that a sample needs to be removed from the pool

Also investigate: what criteria are giving red QC-flag now? We got red flag for pool D1 and F1 in tga200423 but couldn't understand why (https://clinical-lims.scilifelab.se/clarity/work-complete/142925)

**Comment by annaengstrom** [comment]
Also add log file to show if and which sample does not have reads (from Set missing reads i Reception control step)

**Comment by annaengstrom** [comment]
Could we possibly have a discussion on the QC flags in this issue?

**Comment by mayabrandi** [comment]
I think we had that discussion orally. Total Volume (ul) > 17 is the qc threshod used in this step. Should it be something else?

**Comment by annaengstrom** [comment]
We used 15 ul as a threshold, but as far as I understood a failed QC flag in a previous step could also make the flag red right?
FYI @annagellerbring @christoscoucoravas

**Comment by mayabrandi** [comment]
A failed QC falg in previous step can not make the qc flagg in this step red. 
The Set QC  script that is run in the step gives a red flag if Total Volume (ul) > 17
The calculate volumes script also stets a red qc flag if Amount taken (ng) <187.5

Maybe you want to take away the Set QC?
I think best we talk about it in person!

**Comment by annaengstrom** [comment]
I think we should only have a red flag if Amount taken (ng) < 187.5 ng, do you agree @annagellerbring and @christoscoucoravas?

**Comment by christoscoucoravas** [comment]
> A failed QC falg in previous step can not make the qc flagg in this step red.
> The Set QC script that is run in the step gives a red flag if Total Volume (ul) > 17
> The calculate volumes script also stets a red qc flag if Amount taken (ng) <187.5
> 
> Maybe you want to take away the Set QC?
> I think best we talk about it in person!

Hello Maya,

Here you write that the Total volume should not exceed 17ul or else QC red flag. 

Is that the total volume of the pool or the volume per sample? 

Because, I think we need a red QC flag, if the volume of required sample exceeds the 17ul (because that is our maximum volume after Tapestation and Quantit) but the maximum volume of the pool can actually be higher than 17ul.

What do you think? @annaengstrom @annagellerbring @mayabrandi

**Comment by mayabrandi** [comment]
Hi. Total Volume (ul) is on the pool!

**Comment by mayabrandi** [comment]
The volume of a sample in the pool will never be higher than 15 ul:

```
            for art in artifacts:
                reads = art.samples[0].udf.get('Reads missing (M)')
                concentration = art.udf.get('Concentration')
                if None in [reads, concentration]:
                    self.missing_udfs += ['Reads missing (M)', 'Concentration']
                    total_volume = None
                    break
                fract_of_pool = reads/float(total_reads)
                amount = pool_size * fract_of_pool
                vol = amount/concentration
                if vol>15:
                    vol=15
                if amount<187.5:
                    pool.qc_flag='FAILED'
                    self.amount_fail = True
                art.udf['Amount taken (ng)'] = amount
                art.udf['Volume of sample (ul)'] = vol
                art.put()            
                total_volume += vol
```

So id say all we need to do is to remove the set qc button.

**Comment by mayabrandi** [comment]
Regarding missing udf on sample, i prefere to make a check in the step where it is supposed to be set. So that the user can correct it directly. I made a script to check for missing reads. Herer is the RP for it. Can you test it for me and approve if okey? We can do it together if you want.

https://github.com/Clinical-Genomics/clinical_EPPs/pull/167

**Comment by cesve** [comment]
> for art in artifacts:
>                 reads = art.samples[0].udf.get('Reads missing (M)')
>                 concentration = art.udf.get('Concentration')
>                 if None in [reads, concentration]:
>                     self.missing_udfs += ['Reads missing (M)', 'Concentration']
>                     total_volume = None
>                     break
>                 fract_of_pool = reads/float(total_reads)
>                 amount = pool_size * fract_of_pool
>                 vol = amount/concentration
>                 if vol>15:
>                     vol=15
>                 if amount<187.5:
>                     pool.qc_flag='FAILED'
>                     self.amount_fail = True
>                 art.udf['Amount taken (ng)'] = amount
>                 art.udf['Volume of sample (ul)'] = vol
>                 art.put()            
>                 total_volume += vol

**Comment by annagellerbring** [comment]
Correction: 
if amount > "total amount" (recevied from library QC protocol)
pool.qc_flag='FAILED'

**Comment by annaengstrom** [comment]
ändra gärna på stage så kan vi testa.

**Comment by annagellerbring** [comment]
to be discussed between MB, AG and CS on Monday @cesve @mayabrandi

**Comment by cesve** [comment]
If possible, can we sort the pools in well order in the pooling map? Like:
Pool #A1
Pool #B1
Pool #C1
We can name the pools differently if that's needed.

Can we also sort the samples in the pools by there well position:
A:1 
B:1
C:1
D:1
E:1
F:1
G:1
H:1
A:2
B:2
C:2 and so on.

**Comment by mayabrandi** [comment]
The requested issues above have been implemented and will be verified in issue #128
